include/xgboost/linear_updater.h

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2018 by Contributors
+ */
+#pragma once
+
+#include <dmlc/registry.h>
+#include <xgboost/base.h>
+#include <xgboost/data.h>
+#include <functional>
+#include <string>
+#include <utility>
+#include <vector>
+#include "../../src/gbm/gblinear_model.h"
+
+namespace xgboost {
+/*!
+ * \brief interface of linear updater
+ */
+class LinearUpdater {
+ public:
+  /*! \brief virtual destructor */
+  virtual ~LinearUpdater() {}
+  /*!
+   * \brief Initialize the updater with given arguments.
+   * \param args arguments to the objective function.
+   */
+  virtual void Init(
+      const std::vector<std::pair<std::string, std::string> >& args) = 0;
+
+  /**
+   * \fn  virtual void LinearUpdater::Update(std::vector<bst_gpair>* in_gpair, DMatrix* data, gbm::GBLinearModel* model, double sum_instance_weight) = 0;
+   *
+   * \brief Updates linear model given gradients.
+   *
+   * \param in_gpair            The gradient pair statistics of the data.
+   * \param data                Input data matrix.
+   * \param model               Model to be updated.
+   * \param sum_instance_weight The sum instance weight.
```

Comment:
It took me a while to understand why the sum instance weight had to be part of the generic interface. Why don't you copy the comment line 27 of the file `coordinate_common.h`, which says
```
* \param sum_instance_weight the sum instance weight, used to normalise L1/L2 penalty
```
You should also add to `parameter.md` a clarification of L1/L2 penalty: that they will be normalized by the number of training instances, to follow the convention of sklearn/glmnet.

Reply:
Sure, can do. I was thinking of creating a new docs page with info about linear model implementations as a few people have been asking about it.. Perhaps that would be a good place to clarify this.

Reply:
Along with it, can we also make a demo as well?

Reply:
Good point. I agree with your suggestion about having a parameter for the legacy behavior.